### PR TITLE
antehandler refactor

### PR DESCRIPTION
**Pull request by ramacarlucho**
Closes: #XXX

## Description

Refactor ante handler gas functions to allow calling from external modules
______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/ante/eth.go [hunk 1]
```diff
@@ -103,6 +103,29 @@ func NewEthAccountVerificationDecorator(ak AccountKeeper, bankKeeper BankKeeper,
 	}
 }
 
+// ValidateSenderHasEnoughFounds validates sender has enough funds to pay for tx cost
+func ValidateSenderHasEnoughFounds(
```
Comment (fedekunze) [inline app/ante/eth.go:107]: rename to `CheckSenderBalance` and move to a new file `x/evm/keeper/utils.go`

File: app/ante/eth.go [hunk 2]
```diff
@@ -154,18 +177,9 @@ func (avd EthAccountVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx
...
-		if balance.Amount.BigInt().Cmp(cost) < 0 {
-			return ctx, stacktrace.Propagate(
-				sdkerrors.Wrapf(
-					sdkerrors.ErrInsufficientFunds,
-					"sender balance < tx cost (%s < %s%s)", balance, txData.Cost(), evmDenom,
-				),
-				"sender should have had enough funds to pay for tx cost = fee + amount (%s = %s + %s)", cost, txData.Fee(), txData.GetValue(),
-			)
+		err = ValidateSenderHasEnoughFounds(ctx, avd.bankKeeper, from, txData, evmDenom)
+		if err != nil {
```
Comment (fedekunze) [inline app/ante/eth.go:181]: I suggest you use this pattern for errors
```go
		if err := func(); err != nil {
```

File: app/ante/eth.go [hunk 3]
```diff
@@ -274,6 +288,7 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 	blockHeight := big.NewInt(ctx.BlockHeight())
 	homestead := ethCfg.IsHomestead(blockHeight)
 	istanbul := ethCfg.IsIstanbul(blockHeight)
+	evmDenom := egcd.evmKeeper.GetParams(ctx).EvmDenom
```
Comment (fedekunze) [inline app/ante/eth.go:291]:
```suggestion
	evmDenom := params.EvmDenom
```

File: app/ante/eth.go [hunk 4]
```diff
@@ -289,45 +304,9 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
...
-
-		gasLimit := txData.GetGas()
-
-		var accessList ethtypes.AccessList
-		if txData.GetAccessList() != nil {
-			accessList = txData.GetAccessList()
-		}
-
-		intrinsicGas, err := core.IntrinsicGas(txData.GetData(), accessList, isContractCreation, homestead, istanbul)
+		err = DeductTxCostsFromUserBalance(ctx, egcd.bankKeeper, egcd.ak, *msgEthTx, txData, evmDenom, homestead, istanbul)
```
Comment (fedekunze) [inline app/ante/eth.go:307]: ditto error style

File: app/ante/eth.go [hunk 5]
```diff
@@ -348,6 +327,59 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 	return next(ctx, tx, simulate)
 }
 
+// DeductTxCostsFromUserBalance it calculates the tx costs and deducts the fees
+func DeductTxCostsFromUserBalance(
```
Comment (fedekunze) [inline app/ante/eth.go:331]: ditto move to keeper

**Review by fedekunze — COMMENTED** [review]
Thanks @ramacarlucho! Left minor comments. Can you also write a test for the 2 functions?

**Comment by ramacarlucho** [comment]
@fedekunze also moved AccountKeeper, and BankKeeper interfaces to keeper/utils.go to avoid  cyclical dependencies

File: x/evm/keeper/utils.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	authante.AccountKeeper
+	NewAccountWithAddress(ctx sdk.Context, addr sdk.AccAddress) authtypes.AccountI
+	GetSequence(sdk.Context, sdk.AccAddress) (uint64, error)
+}
+
+// BankKeeper defines an expected keeper interface for the bank module's Keeper
+type BankKeeper interface {
+	authtypes.BankKeeper
+	GetBalance(ctx sdk.Context, addr sdk.AccAddress, denom string) sdk.Coin
+}
```
Comment (fedekunze) [inline x/evm/keeper/utils.go:27]: these are already defined in `types/interface.go` we can merge the 2 probably and use those

File: app/ante/eth.go [hunk 7]
```diff
@@ -289,45 +281,8 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
...
-		fees := sdk.Coins{sdk.NewCoin(evmDenom, sdk.NewIntFromBigInt(feeAmt))}
-
-		// deduct the full gas cost from the user balance
-		if err := authante.DeductFees(egcd.bankKeeper, ctx, signerAcc, fees); err != nil {
-			return ctx, stacktrace.Propagate(
-				err,
-				"failed to deduct full gas cost %s from the user %s balance", fees, msgEthTx.From,
-			)
+		if err := evmkeeper.DeductTxCostsFromUserBalance(ctx, egcd.bankKeeper, egcd.ak, *msgEthTx, txData, evmDenom, homestead, istanbul); err != nil {
+			return ctx, err
```
Comment (fedekunze) [inline app/ante/eth.go:285]: use `stacktrace.Propagate`

File: app/ante/eth.go [hunk 8]
```diff
@@ -154,18 +155,8 @@ func (avd EthAccountVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx
...
-		if balance.Amount.BigInt().Cmp(cost) < 0 {
-			return ctx, stacktrace.Propagate(
-				sdkerrors.Wrapf(
-					sdkerrors.ErrInsufficientFunds,
-					"sender balance < tx cost (%s < %s%s)", balance, txData.Cost(), evmDenom,
-				),
-				"sender should have had enough funds to pay for tx cost = fee + amount (%s = %s + %s)", cost, txData.Fee(), txData.GetValue(),
-			)
+		if err := evmkeeper.CheckSenderBalance(ctx, avd.bankKeeper, from, txData, evmDenom); err != nil {
+			return ctx, err
```
Comment (fedekunze) [inline app/ante/eth.go:159]: `stacktrace.Propagate`

**Review by fedekunze — APPROVED** [review]
Very clean, and the table tests look good. Thanks! Left a few minor comments that we can tackle before merging

File: x/evm/keeper/utils.go (hunk 6 above)
Comment (hanchon) [reply x/evm/keeper/utils.go]: AccountKeeper's `GetParams` was the only method missing in the `interfaces.go` file. I added it and now every file is now using the same interfaces.

**Comment by hanchon** [comment]
`DeductTxCostsFromUserBalance` now returns `fee, error` because after merging master the `ante/eth.go` file was modified and the `fee` value is need for the `event` created after deducting the tx costs.
```
events = append(events, sdk.NewEvent(sdk.EventTypeTx, sdk.NewAttribute(sdk.AttributeKeyFee, fees.String())))
```

File: x/evm/keeper/utils.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	if balance.Amount.BigInt().Cmp(cost) < 0 {
+		return stacktrace.Propagate(
+			sdkerrors.Wrapf(
+				sdkerrors.ErrInsufficientFunds,
+				"sender balance < tx cost (%s < %s%s)", balance, txData.Cost(), denom,
+			),
+			"sender should have had enough funds to pay for tx cost = fee + amount (%s = %s + %s)",
+			cost,
+			txData.Fee(),
+			txData.GetValue(),
```
Comment (fedekunze) [inline x/evm/keeper/utils.go:95]:
```suggestion
			cost, txData.Fee(), txData.GetValue(),
```

**Pull request merged**